creating a Locatable interface

lbergelson opened the pull request:
Renaming htsjdk.samtools.util.Interval -> NamedInterval which now extends the minimal Interval
extracting out a minimal genomic interval class from and naming that Interval
SAMRecord implements HasGenomicInterval
Feature implements HasGenomicInterval
renaming classes that are based on NamedInterval
renaming htsjdk.tribble.index.interval.Interval to FileInterval

This is motivated by the need to have a common interface to work with tribble `Feature` and `SAMRecord` for location based caching purposes. 
It's also a first step into integrating the the many interval classes that htsjdk, picard,  and GATK have and pushing functionality into a common base class.  See https://github.com/broadinstitute/hellbender/issues/100   and https://github.com/broadinstitute/hellbender/issues/159

@droazen please review

lbergelson commented:
After continued discussion it seems like changing this interface to Locatable  and pushing the getContig(), getStart(), getEnd() calls into this interface would be a good idea.

mattsooknah commented:
Would it be difficult to have more things use the base Interval instead of requiring a NamedInterval?

It seems like many things don't really care about the name, and it's awkward to keep creating new NamedIntervals with ```name=null```.

lbergelson commented:
Ideally everything that can would be transitioned to `Interval`.  It's some work.  Methods have to be pulled up to `Interval` and then overridden again in `NamedInterval` in a number of cases.  We have a very similar issue in hellbender.

nh13 commented:
What's the impact on Picard?  We have to make sure changes in htsjdk to support internal development effort for (currently) non-production software does not impact production software.

mattsooknah commented:
```Interval``` is used all over the place in Picard. If it's just renaming ```Interval``` -> ```NamedInterval```, then it's a lot of code change but very little effort. But if we wanted to do it right (as Louis described above), then it could get pretty messy...

lbergelson commented:
After getting feedback I've minimized the changes.  This should cause no issues with dependant code.

There is now a new `Locatable` interface which defines a single method ``SimpleInterval getInterval()`
Introduced the `SimpleInterval` which is just a triplet of contig,start,stop.  

`Feature`,`Interval`, and `SAMRecord` now implement `Locatable` for a unified (if slightly awkward) way of handling them.

@mattsooknah @nh13 @droazen Please re-review when you get a change.

nh13 commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -1179,6 +1181,15 @@ protected SAMBinaryTagAndValue getBinaryAttributes() {
         return mAttributes;
     }
 
+    @Override
```
Comment about what this concretely returns.  Does the clipping affect mapping, is it the span over the genome, what happens for unmapped reads etc.?

nh13 commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -1179,6 +1181,15 @@ protected SAMBinaryTagAndValue getBinaryAttributes() {
 
+    @Override
+    public SimpleInterval getInterval() {
+        if (mReferenceName != null && mAlignmentStart > 0){
```
What happens to unmapped reads with a reference and start but no alignment end?  I.e. they are given the start coordinate of their mapped mate.

lbergelson replied:
Changing to getReadUnmappedFlag() instead of this check.

nh13 commented on src/java/htsjdk/samtools/util/Interval.java:
```diff
@@ -63,24 +60,15 @@ public Interval(final String sequence, final int start, final int end) {
+    /** Gets the name of the sequence on which the interval resides.
+     * This is a simple alias of getContig()
+     */
+    public String getSequence() { return getContig(); }
```
should we deprecate since it seems we want to use ```getContig```?

nh13 commented on src/java/htsjdk/samtools/util/SimpleInterval.java:
```diff
@@ -0,0 +1,45 @@
+    public final SimpleInterval getInterval(){
+        return this;
+    }
+
```
remember when I said I would stop being nitpicky? I lied, extra newline.

nh13 commented on src/java/htsjdk/samtools/util/SimpleInterval.java:
```diff
@@ -0,0 +1,45 @@
+    public final int getStart(){
+        return start;
+    }
+    public final int getEnd(){
```
comment, and add some newlines

nh13 commented on src/java/htsjdk/tribble/bed/SimpleBEDFeature.java:
```diff
@@ -136,4 +137,9 @@ public void setLink(String link) {
         return emptyExonList;
     }
+
+    @Override
```
```SimpleInterval``` is 1-based, are all the the classes that implement ```SimpleInterval``` as well?

droazen replied:
I think that internally we should always convert to 1-based closed, and only convert to other representations when writing formats that require it (like BED files).

nh13 replied:
Agreed.  My intuition is that some of the classes that now implement ```getInterval``` are not 1-based closed, and so those method need to be updated.

nh13 commented on src/java/htsjdk/samtools/util/SimpleInterval.java:
```diff
@@ -0,0 +1,45 @@
+
+
+/**
+ * minimal immutable class representing a 1-based closed ended genomic interval
```
thank-you for the clear definition

nh13 commented on src/java/htsjdk/samtools/util/Locatable.java:
```diff
@@ -0,0 +1,9 @@
+     * gets a {@link SimpleInterval} matching the mapping of this locatable to the reference genome
+     * @return a SimpleInterval or null if this is unmapped
+     */
+    SimpleInterval getInterval();
```
Maybe call this ```getSimpleInterval``` since we have a class called ```Interval``` which presumably could be the inferred return type?

droazen replied:
How about `getLocation()` for consistency with the interface name?

nh13 commented on src/java/htsjdk/tribble/BasicFeature.java:
```diff
@@ -24,6 +24,8 @@
+
 /**
  * Trivial feature -- all it holds is the key information.
  *
```
wouldn't it be nice if BasicFeature told you if it was 1-based closed ended, or is it supposed to be undefined?  If the former, document, if the latter, update the ```getInterval``` function.

nh13 commented on src/java/htsjdk/tribble/bed/SimpleBEDFeature.java:
```diff
@@ -23,6 +23,7 @@
+import htsjdk.samtools.util.SimpleInterval;
 import htsjdk.tribble.annotation.Strand;
 
 import java.awt.*;
```
Comment about if a BED is 1-based closed ended (see ```getInterval```).

nh13 commented on src/java/htsjdk/tribble/dbsnp/OldDbSNPFeature.java:
```diff
@@ -23,6 +23,7 @@
+import htsjdk.samtools.util.SimpleInterval;
 import htsjdk.tribble.Feature;
 import htsjdk.tribble.annotation.Strand;
 
```
Comment about if a BED is 1-based closed ended (see ```getInterval```).

nh13 commented on src/java/htsjdk/tribble/gelitext/GeliTextFeature.java:
```diff
@@ -23,6 +23,7 @@
 
+import htsjdk.samtools.util.SimpleInterval;
 import htsjdk.tribble.Feature;
 
```
Comment about if a BED is 1-based closed ended (see ```getInterval```).

nh13 commented on src/java/htsjdk/variant/variantcontext/VariantContext.java:
```diff
@@ -25,6 +25,7 @@
 
 package htsjdk.variant.variantcontext;
 
```
Comment about if a BED is 1-based closed ended (see ```getInterval```).

droazen commented on src/java/htsjdk/samtools/util/SimpleInterval.java:
```diff
@@ -0,0 +1,45 @@
+    private final int end;
+    private final String contig;
+
+    public SimpleInterval(final String contig, final int start, final int end){
```
Add javadoc to constructor

droazen commented on src/java/htsjdk/samtools/util/SimpleInterval.java:
```diff
@@ -0,0 +1,45 @@
+
+    public SimpleInterval(final String contig, final int start, final int end){
+        if(contig == null){
+            throw new IllegalArgumentException("contig cannot be null");
```
Can the current implementation of `SimpleInterval` represent an unmapped interval? If so, how?

nh13 replied:
What is an unmapped interval?

droazen replied:
I know it sounds like an oxymoron, but we actually have this concept in the GATK. Eg., if the user passes in `-L unmapped`, the GATK will create an interval object (a `GenomeLoc`, in our case) to represent this "unmapped" interval, and the engine, when given this unmapped interval object, will then set up a traversal over the unmapped reads in the input file(s). 

If a `SamRecord` can be unmapped, yet implements `Locatable`, doesn't that imply that "Locatables" in general need some way of indicating the absence of a location (ie., something like an `isUnmapped()` method)?

nh13 replied:
It's not an absence of location, it's a location not in the reference.  ```Locatable``` is relative to the reference implicitly, so you can say that it is "not in the reference".  I would not use "unmapped" here since that is confusing to a process (i.e. mapping).  I would ```inLocation``` or ```inReference``` or some positive term (I don't like these terms).

droazen replied:
It is the absence of a *known* location, isn't it?

nh13 replied:
Locatable implies you can know.

droazen replied:
Now we are getting philosophical... to be clear, I am fine with Louis's solution of having `SAMRecord.getInterval()` return null for unmapped reads rather than providing a way for `SimpleInterval` itself to represent the "location unknown" state. I just wanted to bring up the old GATK approach of having `isUnmapped()` built into the `GenomeLoc` class to make sure that we don't want that here.

mattsooknah commented:
If we're going to represent everything as 1-based closed, then presumably zero-length intervals will not be allowed, since they can't be represented. So, e.g. if we read in a BED file with a zero-length interval, we should just ignore it.

(currently, Picard tests are throwing an error on this because they assumes you can represent a zero-length BED feature as an Interval with end = start-1 ... ack!)

yfarjoun commented:
picard interval lists are 1 based closed...zero sized intervals are allowed
and possible. the way to do that is to have the end position be one _less_
than the start position.




On Fri, Feb 27, 2015 at 4:32 PM, Matt Sooknah <notifications@github.com>
wrote:

> If we're going to represent everything as 1-based closed, then presumably
> zero-length intervals will not be allowed, since they can't be represented.
> So, e.g. if we read in a BED file with a zero-length interval, we should
> just ignore it.
>
> (currently, Picard tests are throwing an error on this because they
> assumes you can represent a zero-length BED feature as an Interval with end
> = start-1 ... ack!)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/samtools/htsjdk/pull/182#issuecomment-76475750>.
>

mattsooknah commented:
Yes, what I'm saying is that seems like is a bug, not a feature...

yfarjoun commented:
it is true that GATK intervals cannot represent this...this is one of the
differences between the two engines and it should be resolved by the
community.



On Fri, Feb 27, 2015 at 4:33 PM, Yossi Farjoun <farjoun@broadinstitute.org>
wrote:

> picard interval lists are 1 based closed...zero sized intervals are
> allowed and possible. the way to do that is to have the end position be one
> _less_ than the start position.
>
>
>
>
> On Fri, Feb 27, 2015 at 4:32 PM, Matt Sooknah <notifications@github.com>
> wrote:
>
>> If we're going to represent everything as 1-based closed, then presumably
>> zero-length intervals will not be allowed, since they can't be represented.
>> So, e.g. if we read in a BED file with a zero-length interval, we should
>> just ignore it.
>>
>> (currently, Picard tests are throwing an error on this because they
>> assumes you can represent a zero-length BED feature as an Interval with end
>> = start-1 ... ack!)
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/samtools/htsjdk/pull/182#issuecomment-76475750>.
>>
>
>

yfarjoun commented:
Oh, no...it's a feature!!

On Fri, Feb 27, 2015 at 4:35 PM, Matt Sooknah <notifications@github.com>
wrote:

> Yes, what I'm saying is that seems like is a bug, not a feature...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/samtools/htsjdk/pull/182#issuecomment-76476186>.
>

yfarjoun commented:
for example, the current ICE exome target contains zero-sized intervals.

On Fri, Feb 27, 2015 at 4:36 PM, Yossi Farjoun <farjoun@broadinstitute.org>
wrote:

> Oh, no...it's a feature!!
>
> On Fri, Feb 27, 2015 at 4:35 PM, Matt Sooknah <notifications@github.com>
> wrote:
>
>> Yes, what I'm saying is that seems like is a bug, not a feature...
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/samtools/htsjdk/pull/182#issuecomment-76476186>.
>>
>
>

mattsooknah commented:
why, though?

nh13 commented:
Agreed, lets keep it in.  There are target captures that have zero intervals (this is public, no internal leakage please).

nh13 commented:
insertions

mattsooknah commented:
Ah, I see. In that case, ```SimpleInterval``` will need to be changed to not throw an exception if end = start -1

lbergelson commented:
If 0 length intervals are necessary we need to change to an open ended
interval.  Having an interval with end < start is crazy.

droazen commented:
Given the sheer volume of code in the GATK and elsewhere that assumes 1-based closed intervals, I don't think that switching to an open-ended representation is a viable option, whatever the benefits from the standpoint of mathematical purity.

lbergelson commented:
@droazen updated

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -1180,6 +1182,33 @@ protected SAMBinaryTagAndValue getBinaryAttributes() {
+
+    @Override
+    public String getContig() {
+        return null;
```
Why is this not returning `mReferenceName`?

lbergelson replied:
because it's completely broken... huh, not sure how I managed that one.

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -1180,6 +1182,33 @@ protected SAMBinaryTagAndValue getBinaryAttributes() {
+
+    @Override
+    public int getEnd() {
+        return 0;
```
Why are these not returning `mAlignmentStart` / mAlignmentEnd` ?

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -1180,6 +1182,40 @@ protected SAMBinaryTagAndValue getBinaryAttributes() {
+
+    /**
+     * @return get
+     */
```
Delete this incomplete javadoc

droazen commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -1180,6 +1182,40 @@ protected SAMBinaryTagAndValue getBinaryAttributes() {
+        if( getReadUnmappedFlag()) {
+            return null;
+        } else {
+            return getReferenceName();
```
This seems fine given the disclaimer in the comments for this class that `Only if the readUnmappedFlag is false can the reference name/index and alignment start be interpreted as indicating an actual alignment position.`

droazen commented on src/java/htsjdk/samtools/util/SimpleInterval.java:
```diff
@@ -0,0 +1,54 @@
+     */
+    public SimpleInterval(final String contig, final int start, final int end){
+        if(contig == null){
+            throw new IllegalArgumentException("contig cannot be null");
```
`Locatable` interface documentation allows null contig names, though:

     /**
      * Gets the contig name for the contig this is mapped to.  May return null if there is no unique mapping.
      * @return name of the contig this is mapped to, potentially null
      */
     String getContig();

lbergelson replied:
Yes, because some `Locatable` might be unmapped.  A `SimpleInterval` is always mapped or it is meaningless.  We can relax that for simple interval if you think it's weird.

droazen replied:
No, that's fine -- but you should explicitly mention this in the class documentation

lbergelson replied:
done

droazen commented on src/java/htsjdk/tribble/BasicFeature.java:
```diff
@@ -24,31 +24,20 @@
-public class BasicFeature implements Feature {
-    private final String chr;
-    private final int start, stop;
+public class BasicFeature extends SimpleInterval implements Feature {
```
Since this is a `Feature`, it shouldn't extend `SimpleInterval`

lbergelson replied:
aaaaah

lbergelson replied:
changed.  these extra 12 lines of autogenerated code are on your head!

droazen replied:
I can live with it :)

droazen commented on src/java/htsjdk/tribble/Feature.java:
```diff
@@ -24,24 +24,18 @@
+     * @deprecated use getContig() instead
      */
+    @Deprecated
     public String getChr();
```
Create separate ticket to remove calls to the now-deprecated `getChr()` in the GATK as well

lbergelson replied:
see https://github.com/broadinstitute/hellbender/issues/259

droazen commented on src/java/htsjdk/tribble/dbsnp/OldDbSNPFeature.java:
```diff
@@ -82,10 +85,16 @@
      */
 
+    @Deprecated
     public String getChr() {
```
Should be @Override

lbergelson replied:
done

droazen commented on src/java/htsjdk/variant/variantcontext/VariantContext.java:
```diff
@@ -1605,14 +1606,32 @@ private final Genotype fullyDecodeGenotypes(final Genotype g, final VCFHeader he
     // tribble integration routines -- not for public consumption
     //
     // ---------------------------------------------------------------------------------------------------------
+    @Deprecated
```
@Override as well

lbergelson replied:
done

droazen commented on src/java/htsjdk/variant/variantcontext/VariantContext.java:
```diff
@@ -1605,14 +1606,32 @@ private final Genotype fullyDecodeGenotypes(final Genotype g, final VCFHeader he
 
+    /**
+     * @return 1-based start position of the Variant,
+     * INDELS events usually start on the first unaltered reference base before the INDEL
```
INDELS -> INDEL

lbergelson replied:
done

droazen commented on src/java/htsjdk/variant/variantcontext/VariantContext.java:
```diff
@@ -1605,14 +1606,32 @@ private final Genotype fullyDecodeGenotypes(final Genotype g, final VCFHeader he
     }
 
+    /**
+     * @return 1-based start position of the Variant,
```
1-based -> 1-based inclusive

lbergelson replied:
done

droazen commented on src/java/htsjdk/variant/variantcontext/VariantContext.java:
```diff
@@ -1605,14 +1606,32 @@ private final Genotype fullyDecodeGenotypes(final Genotype g, final VCFHeader he
     }
 
+    /**
+     * @return 1-based closed end position of the Variant, if the END info field is specified that value is returned
```
should note: otherwise the end is the start + reference allele length - 1

lbergelson replied:
should we remove the stuff about indels?

droazen commented on src/tests/java/htsjdk/samtools/util/SimpleIntervalTest.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Test
+    public void testZeroLengthInterval(){
+        SimpleInterval interval = new SimpleInterval("1",100,99);
```
Call `getStart()` and `getEnd()` in this test as well

lbergelson replied:
done

droazen commented:
:+1: new version looks good -- just a few minor comments here and there

lbergelson commented:
@nh13 could you take a look at this again?

nh13 commented on src/java/htsjdk/samtools/SAMRecord.java:
```diff
@@ -1180,6 +1182,48 @@ protected SAMBinaryTagAndValue getBinaryAttributes() {
+        if (getReadUnmappedFlag()) {
+            return null;
+        } else {
+            return new SimpleInterval(mReferenceName, mAlignmentStart, mAlignmentEnd);
```
This should probably use ```getAlignmentEnd``` instead of ```mAlignmentEnd```, since the later may be unset for mapped reads until the former is called.  It is probably better practice to use the getters in these calls anyhow.

lbergelson replied:
done

nh13 commented on src/java/htsjdk/tribble/bed/BEDFeature.java:
```diff
@@ -31,6 +31,10 @@
  * @date Dec 24, 2009
+ *
+ * BED feature start and end positions must adhere to the Feature interval specifications
+ * This is different than the 0-based representation in a BED file.  This conversion is handled by {@link BEDCodec}
```
periods in lines 35/36.

lbergelson replied:
done

nh13 commented on src/java/htsjdk/variant/variantcontext/VariantContext.java:
```diff
@@ -1605,14 +1606,33 @@ private final Genotype fullyDecodeGenotypes(final Genotype g, final VCFHeader he
     }
 
+    /**
+     * @return 1-based inclusive start position of the Variant,
```
dangling comma

lbergelson replied:
done

nh13 commented on src/java/htsjdk/variant/variantcontext/VariantContext.java:
```diff
@@ -1605,14 +1606,33 @@ private final Genotype fullyDecodeGenotypes(final Genotype g, final VCFHeader he
+     * the start + reference allele length - 1
+     * insertions end on the reference base before the insertion event
+     * deletions end on the final deleted reference base
+     */
```
The formatting of this comment is confusing (no periods, newlines, capitalization).

lbergelson replied:
improved the comment

nh13 commented:
There are a few minor comments, please fix, and then :+1:

lbergelson commented:
Squashed commits

I also added `@Deprecated` to the other 2 members of the `gelitext` package since the codec is now deprecated and no else uses those classes

lbergelson commented:
@nh13 I think it's ready to merge unless there are any other comments.

nh13 commented:
@mattsooknah can you verify that this version will work with our Picard?

mattsooknah commented:
All tests pass. S'good to go.

lbergelson commented:
@nh13 At the request of @droazen I've added an equals and hashcode method to SimpleInterval.  Should be a non breaking change.  Going to merge on a thumbs up and travis completion.

lbergelson commented:
Also a SimpleInterval.toString().  Getting crazy over here.

lbergelson commented:
@nh13 Good to merge?

nh13 commented:
Nice work!  Go for it.
